src/util/mpscfifo.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <QAtomicInteger>
+
+#include "util/assert.h"
+
+// Lock-free FIFO for multiple producers/writers and a single(!) consumer/reader.
+template<typename T, int capacity>
+class MpscFifo {
+  public:
+    MpscFifo()
+          : m_enqueueSize(0),
+            m_dequeueSize(0),
+            m_headIndex(0),
+            m_tailIndex(0) {
+        static_assert(capacity >= 2, "capacity too low");
+    }
+
+    bool enqueue(T value) {
+        if (m_enqueueSize.fetchAndAddAcquire(1) >= capacity) {
+            // Queue is full -> Restore size and abort
+            m_enqueueSize.fetchAndSubRelease(1);
+            return false;
+        } else {
+            int headIndex = m_headIndex.fetchAndAddAcquire(1);
+            DEBUG_ASSERT(headIndex >= 0);
+            m_buffer[headIndex % capacity] = std::move(value);
+            // Allow the reader to access the enqueued value
+            m_dequeueSize.fetchAndAddRelease(1);
+            return true;
+        }
+    }
+
+    bool dequeue(T* value) {
+        if (m_dequeueSize.fetchAndSubAcquire(1) <= 0) {
+            // Queue is empty -> Restore size and abort
+            m_dequeueSize.fetchAndAddRelease(1);
+            return false;
+        } else {
+            DEBUG_ASSERT(m_tailIndex >= 0);
+            DEBUG_ASSERT(m_tailIndex < capacity);
+            *value = std::move(m_buffer[m_tailIndex]);
+            if (++m_tailIndex >= capacity) {
+                m_tailIndex %= capacity;
+                // Prevent unlimited growth and overflow of m_headIndex which
+                // is already ahead of m_tailIndex.
+                int headIndex;
+                do {
+                    headIndex = m_headIndex.load();
+                    DEBUG_ASSERT(headIndex >= 0);
+                } while ((headIndex >= capacity) &&
+                        !m_headIndex.testAndSetOrdered(headIndex, headIndex % capacity));
+            }
+            // Allow the writer to overwrite the dequeued value
+            m_enqueueSize.fetchAndSubRelease(1);
+            return true;
+        }
+    }
+
+  private:
+    T m_buffer[capacity];
+    QAtomicInteger<int> m_enqueueSize;
```

Comment:
Can we use here the non template QAtomicInt here? I think it is not worth to advance the minimum Qt requirent to Qt 5.3 because of that.

Reply:
https://github.com/uklotzde/mixxx/pull/9

Reply:
Done (not yet pushed).

Those ancient system Qt versions add complexity and pile up more legacy code. Times are changing and so must Debian/Ubuntu, but we can't do anything about it. Mixxx must be detached from their LTS release cycles by using Snap/Flatpak/AppImage.